### PR TITLE
Allow float data to be used with our sparse solver extensions

### PR DESCRIPTION
**Pull request by luisenp**
Note that this doesn't changed the extensions themselves. We convert from float to double and back before calling the C++ routines. 

PS: I had to make the same change in the two different solvers, so I really think it's worth unifying the sparse solvers under a common base class. I'll try to find time to do this soon.

### CODE REVIEW

**Review by maurimo — APPROVED** [review]
LGTM!

File: tests/optimizer/autograd/common.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+    A_val, b = inputs[0], inputs[1]
+    # Check that the gradient works correctly for floating point data
+    out = solve_func(*inputs).sum()
+    gA, gb = grad(out, (A_val, b))
+
+    A_float = A_val.float()
+    b_float = b.float()
+    inputs2 = (A_float, b_float) + inputs[2:]
+    print(type(inputs2), len(inputs2))
```
[the remark this answers is not in the document]
Comment (luisenp) [reply tests/optimizer/autograd/common.py:16]: Oops, will remove.

File: tests/optimizer/autograd/common.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import torch
```
Comment (mhmukadam) [inline tests/optimizer/autograd/common.py:1]: Not sure if this was added later but this file is missing license header in this PR.

File: tests/optimizer/autograd/common.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    gA, gb = grad(out, (A_val, b))
+
+    A_float = A_val.float()
+    b_float = b.float()
+    inputs2 = (A_float, b_float) + inputs[2:]
+    print(type(inputs2), len(inputs2))
+    out_float = solve_func(*inputs2).sum()
+    gA_float, gb_float = grad(out_float, (A_float, b_float))
+
+    torch.testing.assert_close(gA, gA_float.double(), rtol=rtol, atol=atol)
```
[the remark this answers is not in the document]
Comment (mhmukadam) [reply tests/optimizer/autograd/common.py:29]: The question and comments both make sense. I guess if this shows up frequently in practice (for us or users) then we may want to have an option to turn this off (or keep off by default aside from some unit tests with known behavior).

**Review by mhmukadam — COMMENTED** [review]
LGTM!

**Pull request merged**